rust_src/src/floatfns.rs

```diff
@@ -26,3 +52,167 @@ pub fn fmod_float(x: LispObject, y: LispObject) -> LispObject {
+        (
+            $(
+                $required_args: LispObject
+            ),*
+
+            // optional arguments are checked for NILP then turned into Option<LispObject>
+            $(
+                , $optional_args: LispObject
+            )*
+        ) -> LispObject
+        {
+            $(
+                let $optional_args = if NILP($optional_args) {
+                    None
+                } else {
+                    Some($optional_args)
+                };
+            );*
+
+            $($body);*
+        }
+
+        lazy_static! {
+            pub static ref $lisp_struct_name: LispSubr = LispSubr {
+                header: VectorLikeHeader {
+                    size: ((PvecType::PVEC_SUBR as libc::c_int) <<
+                           PSEUDOVECTOR_AREA_BITS) as ptrdiff_t,
+                },
+                function: ($rust_name as *const libc::c_void),
+                // include a trash string so the compiler can derive a type for the vector
+                min_args: (vec![ "trash", $(stringify!($required_args)),* ].len() - 1) as i16,
+                max_args: (
+                    vec![ "trash", $(stringify!($required_args)),* ].len() - 1
+                    + vec![ "trash", $(stringify!($optional_args)),* ].len() - 1) as i16,
+                symbol_name: ((concat!($lisp_name, "\0")).as_ptr()) as *const c_char,
+                intspec: ptr::null(),
+                doc: (concat!($lisp_docs, "\0").as_ptr()) as *const c_char
+            };
+        }
+    }
+}
+
+macro_rules! simple_float_op {
+    ($lisp_name:expr, $rust_name:ident, $sname:ident, $float_func:ident, $lisp_docs:expr) => {
+        lisp_fun!(
+            $rust_name(x) {
+                let d = unsafe { extract_float(x) };
+                let val = d.$float_func();
+                unsafe { make_float(val) }
+            }
+
+            lisp_name: $lisp_name;
+            struct_name: $sname;
+            doc: $lisp_docs;
+        );
+    }
+}
+
+simple_float_op!("acos", Facos, Sacos, acos, "Return the inverse cosine of ARG.");
+simple_float_op!("asin", Fasin, Sasin, asin, "Return the inverse sine of ARG.");
+// atan is special, defined later
+simple_float_op!("cos",  Fcos,  Scos,  cos, "Return the cosine of ARG.");
+simple_float_op!("sin",  Fsin,  Ssin,  sin, "Return the sine of ARG.");
+simple_float_op!("tan",  Ftan,  Stan,  tan, "Return the tangent of ARG.");
+
+simple_float_op!("exp",  Fexp,  Sexp,  exp,  "Return the exponential base e of ARG.");
+simple_float_op!("sqrt", Fsqrt, Ssqrt, sqrt, "Return the square root of ARG.");
+simple_float_op!("abs",  Fabs,  Sabs,  abs,  "Return the absolute value of ARG.");
+
+simple_float_op!("fceiling", Ffceiling, Sfceiling, ceil,
+    "Return the smallest integer no less than ARG, as a float.
+(Round toward +inf.)");
+
+simple_float_op!("ffloor", Ffloor, Sffloor, floor,
+    "Return the largest integer no greater than ARG, as a float.
+(Round towards -inf.)");
+
+lisp_fun! {
```

Comment:
We have a `defun!` macro now, could you use that instead?

I appreciate it didn't exist when you started on this PR, sorry about that.